[Terra-Clinical-Header] A11y Updates of terra-clinical-header

supreethmr opened the pull request:
### Summary
<!--- Summarize and explain the reason behind these code changes. What are the changes, and why are they necessary? -->
Allows users to be able to select a heading level, so that it does not have accessibility assumptions or defaults that create incorrect context for users and my application meets [WCAG Success Criterion 1.3.1 (Info and Relationships)](https://www.w3.org/TR/WCAG21/#info-and-relationships) plus Success Criterion 1.3.2 (Meaningful Sequence).
Allows users to be able to add id to the header to be used with ‘aria-labelledby’ to associate headings with the corresponding page region​

This PR includes changes from[ #792](https://github.com/cerner/terra-clinical/pull/792) , [#794](https://github.com/cerner/terra-clinical/pull/794)

<!--- Include any issue addressed by this pull request. -->
<!--- Example: Closes #45 -->
Closes # UXPLATFORM-8173

### Deployment Link
<!---Include the deployment link, if applicable. -->
<!--- Example: https://terra-clinical-deployed-pr-45.herokuapp.com/ -->
https://terra-clinical-deployed-pr-#.herokuapp.com/

### Testing
<!-- Demonstrate that these changes are stable. How have these changes been verified? -->
Tested with VO and Jaws.

### Additional Details
<!-- List anything else that is relevant to this issue. Additional information will help us better understand your changes and speed up the review process. -->

<!--
*Before publishing*

1. Assign yourself to the PR.
2. Add the appropriate labels
3. Add your name to the CONTRIBUTORS.md file. Adding your name to the CONTRIBUTORS.md file signifies agreement to all rights and reservations provided by the License.
-->

Thank you for contributing to Terra.
@cerner/terra

sdadn commented:
Can you rebase with main since this PR has gotten stale?

sdadn commented on packages/terra-clinical-header/src/Header.jsx:
```diff
@@ -21,7 +21,25 @@ const propTypes = {
+
+  /**
+   * ![IMPORTANT](https://badgen.net/badge/prop/deprecated/red)
+   * title prop has been deperecated and will be removed on next major version relase. Replace the `title` prop with `text` prop.
```
```suggestion
   * The `title` prop has been deprecated and will be removed in next major version release. Use the `text` prop instead.
```

sdadn commented on packages/terra-clinical-header/CHANGELOG.md:
```diff
@@ -3,6 +3,9 @@
 ## Unreleased
 
 * Changed
+  * Replaced `title` prop to `text` prop to allow users to spread title HTML Attribute as customProps.
+  * Added `level` prop to allow users to render different heading levels.
+  * Added `id` prop which is to be paired with `aria-labelledby` for labelling page regions.
   * Updated wdio screenshots.
```
```suggestion
* Added

  * Added `level` prop to allow users to render different heading levels.
  * Added `id` prop which is to be paired with `aria-labelledby` for labelling page regions.
  * Added `text` prop as a replacement for the deprecated `title` prop to allow users to utilize the title HTML Attribute as customProps.
  
* Changed

  * Deprecated `title` prop.
```

supreethmr replied:
updated [here](https://github.com/cerner/terra-clinical/pull/823/commits/7cb1cc54d09c6b0f308fdeb4e38f55dd39d88f4c#diff-46d09a951255f1060174456bc00fb8b3aac50ec092250ab3191ea81a008943fa)

sdadn commented on packages/terra-clinical-item-collection/CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## Unreleased
 
 * Changed
+  * Updated jest snapshots terra-icon A11y updatess.
```
We can remove this as it is no longer needed with the dangerfile changes.

supreethmr replied:
removed [here](https://github.com/cerner/terra-clinical/pull/823/commits/7cb1cc54d09c6b0f308fdeb4e38f55dd39d88f4c#diff-2836fb928f9d253961572dc46bc45f787fc9c2abb9f2b56679f04387c4d80c2e)

sdadn commented on packages/terra-clinical-onset-picker/CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## Unreleased
 
 * Changed
+  * Updated jest snapshots terra-icon A11y updatess.
```
We can remove this as it is no longer needed with the dangerfile changes.

supreethmr replied:
removed [here](https://github.com/cerner/terra-clinical/pull/823/commits/7cb1cc54d09c6b0f308fdeb4e38f55dd39d88f4c#diff-2836fb928f9d253961572dc46bc45f787fc9c2abb9f2b56679f04387c4d80c2e)

sdadn commented on packages/terra-clinical-header/src/Header.jsx:
```diff
@@ -35,27 +53,44 @@ const propTypes = {
+  }
+  if (!level) {
+    // eslint-disable-next-line no-console
+    console.warn('Default heading level may not appropriate has it would fail to convey context of heading in a site / application where it is used. Heading level should be set explicitly depending on the position of header in site / application to allow screen readers to identify headers consistently.'); // to be removed on next major version release.
```
```suggestion
    // TODO: remove on the next major version bump
    console.warn('Default heading level may not appropriate has it would fail to convey context of heading in a site / application where it is used. Heading level should be set explicitly depending on the position of header in site / application to allow screen readers to identify headers consistently.');
```

supreethmr replied:
updated [here](https://github.com/cerner/terra-clinical/pull/823/commits/7cb1cc54d09c6b0f308fdeb4e38f55dd39d88f4c#diff-8c5f8d7501709c71d62d2a51f32a0f10dd712d5a0ede7105f9e0979c4ab845ff)

sdadn commented on packages/terra-clinical-header/src/Header.jsx:
```diff
@@ -35,27 +53,44 @@ const propTypes = {
 
 const Header = ({
-  children, title, startContent, endContent, isSubheader, ...customProps
+  children, title, startContent, endContent, text, level, id, isSubheader, ...customProps
```
Where is `id` being used?

supreethmr replied:
Good Catch !! I missed it assigning to header . 
updated [here](https://github.com/cerner/terra-clinical/pull/823/commits/7cb1cc54d09c6b0f308fdeb4e38f55dd39d88f4c#diff-8c5f8d7501709c71d62d2a51f32a0f10dd712d5a0ede7105f9e0979c4ab845ff)

supreethmr commented:
> Can you rebase with main since this PR has gotten stale?

Done !!

supreethmr commented on packages/terra-clinical-header/src/Header.jsx:
```diff
@@ -35,27 +53,44 @@ const propTypes = {
   const theme = useContext(ThemeContext);
+  if (title) {
+    // eslint-disable-next-line no-console
+    console.warn('`title` prop has been renamed to `text`. please update all the refernces of `title` prop to use prop `text`.'); // to be removed on next major version release.
```
Updated [here](https://github.com/cerner/terra-clinical/pull/823/commits/7cb1cc54d09c6b0f308fdeb4e38f55dd39d88f4c#diff-8c5f8d7501709c71d62d2a51f32a0f10dd712d5a0ede7105f9e0979c4ab845ff)

sdadn commented:
Is this PR still needed?

supreethmr commented:
> Is this PR still needed?

yes !! we lost track of this PR while fixing WDIO snapshot issue. I have rebased the branch with main and updated the WDIO snapshots as well to fix build issue

sdadn commented on packages/terra-clinical-header/CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
 
 ## Unreleased
 
+* Changed
+  * Replaced `title` prop to `text` prop to allow users to spread title HTML Attribute as customProps.
+  * Added `level` prop to allow users to render different heading levels.
+  * Added `id` prop which is to be paired with `aria-labelledby` for labelling page regions.
+
```
```suggestion
* Breaking Changes
  * Renamed `title` prop to `text` prop to allow users to spread title HTML Attribute as customProps.
  
* Added
  * Added `level` prop to allow users to render different heading levels.
  * Added `id` prop which is to be paired with `aria-labelledby` for labelling page regions.

```

sdadn commented on packages/terra-clinical-header/src/Header.jsx:
```diff
@@ -35,34 +53,53 @@ const propTypes = {
    */
   isSubheader: PropTypes.bool,
+
+  /**
+   * IDs should be used with 'aria-labelledby' to associate headings with the corresponding page area when needing to label regions.
+   *
+   * For an example, read [Accessibility Guide: Labeling Section Content](/components/terra-clinical-header/clinical-header/accessibility-guide#labeling-section-content) for additional information.
+   */
+  id: PropTypes.string,
```
Let's move `id` to the top of the list since it's an identifier and can be on top of the props list.
```suggestion
  /**
   * IDs should be used with 'aria-labelledby' to associate headings with the corresponding page area when needing to label regions.
   *
   * For an example, read [Accessibility Guide: Labeling Section Content](/components/terra-clinical-header/clinical-header/accessibility-guide#labeling-section-content) for additional information.
   */
  id: PropTypes.string,
```

sdadn commented on packages/terra-clinical-header/CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
 
 ## Unreleased
 
+* Changed
+  * Replaced `title` prop to `text` prop to allow users to spread title HTML Attribute as customProps.
+  * Added `level` prop to allow users to render different heading levels.
+  * Added `id` prop which is to be paired with `aria-labelledby` for labelling page regions.
+
```
Nit for better clarity of the changes:
```suggestion
* Changed
  * Deprecated `title` prop. This will be replaced by the`text` prop in the next major version.
    
* Added
  * Added `level` prop to allow users to render different heading levels.
  * Added `id` prop which is to be paired with `aria-labelledby` for labelling page regions.
  * Added `text` prop to allow users to spread title HTML Attribute as customProps. This will replace the `title` prop in the next major release.
  
```

supreethmr replied:
updated [changelog](https://github.com/cerner/terra-clinical/pull/823/commits/2a0747c1873dccb3053cc005b59cdea87205589c#diff-46d09a951255f1060174456bc00fb8b3aac50ec092250ab3191ea81a008943fa)

sdadn commented on packages/terra-clinical-header/tests/jest/Header.test.jsx:
```diff
@@ -6,45 +6,52 @@ it('should render a default component', () => {
   expect(subheader).toMatchSnapshot();
 });
+
+it('should render a header with no text from level not set', () => {
+  const header = render(<Header text="This title should not render without a provided level" />);
+  expect(header).toMatchSnapshot();
+});
```
We should explicitly test for the value in addition to the snapshot.

supreethmr replied:
This was added to validate non-passive change on it's previous PR. now with Passive change we still render header with title even when value is not set for prop `level`. It uses default level h1 to maintain passivity.

Updated the test to validate console warning thrown when value for `level` prop is not set https://github.com/cerner/terra-clinical/pull/823/commits/2a0747c1873dccb3053cc005b59cdea87205589c#diff-2a241fab4e9f4e0d99b828bc9f8e3a451b22d15bcb991b77784d45ceb03af1de

sdadn commented on packages/terra-clinical-header/tests/wdio/header-spec.js:
```diff
@@ -52,4 +52,22 @@ Terra.describeViewports('Clinical header', ['tiny', 'small', 'medium', 'large',
     Terra.validates.element('no title with content on right');
   });
+
+  it('Header used to label different regions', () => {
```
```suggestion
  it('uses a heade to label different regions', () => {
```

supreethmr replied:
updated : https://github.com/cerner/terra-clinical/pull/823/commits/2a0747c1873dccb3053cc005b59cdea87205589c#diff-17930a9789595a8fe280df0bac7f10c1ab42f97d088b819afada3036b66b3612

sdadn commented on packages/terra-clinical-header/tests/wdio/header-spec.js:
```diff
@@ -52,4 +52,22 @@ Terra.describeViewports('Clinical header', ['tiny', 'small', 'medium', 'large',
+    Terra.validates.element('labeling regions');
+  });
+
+  it('Header with long title and children', () => {
```
```suggestion
  it('uses a header with long title and children', () => {
```

supreethmr replied:
updated : https://github.com/cerner/terra-clinical/pull/823/commits/2a0747c1873dccb3053cc005b59cdea87205589c#diff-17930a9789595a8fe280df0bac7f10c1ab42f97d088b819afada3036b66b3612

sdadn commented on packages/terra-clinical-header/tests/wdio/header-spec.js:
```diff
@@ -52,4 +52,22 @@ Terra.describeViewports('Clinical header', ['tiny', 'small', 'medium', 'large',
+    Terra.validates.element('long title with children');
+  });
+
+  it('Header with no level and no title', () => {
```
```suggestion
  it('uses a header with no level and no title', () => {
```

supreethmr replied:
updated : https://github.com/cerner/terra-clinical/pull/823/commits/2a0747c1873dccb3053cc005b59cdea87205589c#diff-17930a9789595a8fe280df0bac7f10c1ab42f97d088b819afada3036b66b3612

sdadn commented on packages/terra-clinical-header/src/Header.jsx:
```diff
@@ -35,34 +53,53 @@ const propTypes = {
   const content = React.Children.map(children, child => (
     React.cloneElement(child, {
-      className: classNames([cx('flex-collapse'), children.props.className]),
+      className: classNames([cx('flex-collapse'), child.props.className]),
```
why is `children` changed to `child`?

supreethmr replied:
this was failing when we had multiple nodes / elements wrapped within clinical-header. children becomes array when it has more than one child items in this scenario it fail to retrieve class-names

sdadn commented on packages/terra-clinical-header/src/Header.jsx:
```diff
@@ -35,34 +53,53 @@ const propTypes = {
 
   let titleElement;
-  if (title) {
+  if (title || text) {
```
+1 on maintaining passivity

sdadn commented on packages/terra-clinical-header/src/Header.jsx:
```diff
@@ -35,34 +53,53 @@ const propTypes = {
+  children, title, startContent, endContent, text, level, id, isSubheader, ...customProps
 }) => {
   const theme = useContext(ThemeContext);
+  if (title) {
+    // TODO: remove on the next major version bump
+    // eslint-disable-next-line no-console
+    console.warn('The `title` prop has been renamed to `text`. Please update all references of `title` prop to `text`.');
+  }
+  if (!level) {
+    // TODO: remove on the next major version bump
+    // eslint-disable-next-line no-console
+    console.warn('Default heading level may not appropriate has it would fail to convey context of heading in a site / application where it is used. Heading level should be set explicitly depending on the position of header in site / application to allow screen readers to identify headers consistently.');
+  }
```
We should add unit tests for these to test if console.warn has the appropriate output.

supreethmr replied:
added jest tests : https://github.com/cerner/terra-clinical/pull/823/commits/2a0747c1873dccb3053cc005b59cdea87205589c#diff-17930a9789595a8fe280df0bac7f10c1ab42f97d088b819afada3036b66b3612

supreethmr commented:
@sdadn  I have addressed all the comments. Can you review again !!

sdadn commented on packages/terra-clinical-header/tests/jest/Header.test.jsx:
```diff
@@ -1,50 +1,63 @@
 import React from 'react';
 import Header from '../../src/Header';
 
+const consoleSpy = jest.spyOn(global.console, 'warn');
```
We should utilize the `beforeAll`, `afterAll` and `afterEach` functions so that the mocks are properly restored and reset after each use.

e.g.:

https://github.com/cerner/terra-framework/blob/5b5d493c6e6b122e18c13e670b85e2abdb1c76f5/packages/terra-worklist-data-grid/tests/jest/WorklistDataGrid.test.jsx#L45-L59

supreethmr replied:
[Updated](https://github.com/cerner/terra-clinical/pull/823/commits/a516df0d6cef43cfa1b1b0d6b3978fd479461d1f) : 
with reference to https://jestjs.io/docs/jest-object#jestspyonobject-methodname added `restoreAllMocks()`
